rest/blip_client_test.go

```diff
@@ -65,19 +71,229 @@ type BlipTesterClient struct {
+		for seq := since; seq <= seqLast; seq++ {
+			doc, ok := c.getClientDocForSeq(seq)
+			// filter non-latest entries in cases where we haven't pruned _seqStore
+			if !ok {
+				continue
+			} else if latestDocSeq := doc.latestSeq(); latestDocSeq != seq {
+				// this entry should've been cleaned up from _seqStore
+				base.AssertfCtx(base.TestCtx(c.TB()), "seq %d found in _seqStore but latestSeq for doc %d - this should've been pruned out!", seq, latestDocSeq)
+				continue
+			}
+			if !yield(seq, doc) {
+				c.TB().Logf("OneShotDocsSince: since=%d, _seqLast=%d - stopping iteration", since, seqLast)
+				return
+			}
+		}
+		c.TB().Logf("OneShotDocsSince: since=%d, _seqLast=%d - done", since, seqLast)
+	}
+}
+
+func (c *BlipTesterCollectionClient) docsSince(ctx context.Context, since clientSeq, continuous bool) chan *clientDoc {
+	ch := make(chan *clientDoc)
+	c.goroutineWg.Add(1)
+	go func() {
+		defer c.goroutineWg.Done()
+		sinceVal := since
+		for {
+			if ctx.Err() != nil {
+				close(ch)
+				return
+			}
+			c.TB().Logf("docsSince: sinceVal=%d", sinceVal)
+			for _, doc := range c.OneShotDocsSince(ctx, sinceVal) {
+				select {
+				case <-ctx.Done():
+					close(ch)
+					return
+				case ch <- doc:
+					c.TB().Logf("sent doc %q to changes feed", doc.id)
+					sinceVal = doc.latestSeq()
+				}
+			}
+			if !continuous {
+				c.TB().Logf("opts.Continuous=false, breaking changes loop")
+				break
+			}
+		}
+	}()
+	return ch
+}
+
+type clientSeq uint64
+
+// clientDocRev represents a revision of a document stored on this client, including any metadata associated with this specific revision.
+type clientDocRev struct {
+	clientSeq clientSeq
+	version   DocVersion
+	body      []byte
+	isDelete  bool
+	message   *blip.Message // rev or norev message associated with this revision when replicated
+}
+
+// clientDoc represents a document stored on the client - it may also contain older versions of the document.
+type clientDoc struct {
+	id                   string                     // doc ID
+	lock                 sync.RWMutex               // protects all of the below properties
+	_latestSeq           clientSeq                  // Latest sequence number we have for the doc - the active rev
+	_latestServerVersion DocVersion                 // Latest version we know the server had (via push or a pull)
+	_revisionsBySeq      map[clientSeq]clientDocRev // Full history of doc from client POV
+	_seqsByVersions      map[DocVersion]clientSeq   // Lookup from version into revisionsBySeq
+}
+
+// docRevSeqsNewestToOldest returns a list of sequences associated with this document, ordered newest to oldest.
+// Can be used for lookups in clientDoc.revisionsBySeq
+func (cd *clientDoc) docRevSeqsNewestToOldest() []clientSeq {
+	cd.lock.RLock()
+	defer cd.lock.RUnlock()
+	return cd._docRevSeqsNewestToOldest()
+}
+
+func (cd *clientDoc) _docRevSeqsNewestToOldest() []clientSeq {
+	cd.lock.RLock()
+	defer cd.lock.RUnlock()
```

Comment:
I assume this is a mistake since `docRevSeqsNewestToOldest` is also called with this lock.